
# 🎯 Atividade Prática Final: O Duelo dos Métodos

Agora é sua vez de colocar a mão na massa. Vamos encontrar a raiz da função:
$$ f(x) = e^{-x} - x $$

Sabemos graficamente que a raiz está no intervalo $[0, 1]$.

### Tarefa 1: Bisseção
Use a função `bissecao` que definimos anteriormente para encontrar a raiz no intervalo $[0, 1]$ com tolerância de $10^{-6}$.
* **Pergunta:** Quantas iterações foram necessárias?

### Tarefa 2: Ponto Fixo
Para usar o ponto fixo, podemos isolar o $x$ de duas formas:
1.  **Forma A:** $x = e^{-x} \implies g_1(x) = e^{-x}$
2.  **Forma B:** $e^{-x} = x \implies -x = \ln(x) \implies g_2(x) = -\ln(x)$

Teste o método `ponto_fixo` com a **Forma A** (chute inicial 0.5). Funcionou?
Agora tente aplicar a **Forma B**. O que acontece se você tentar calcular $g_2(0.5)$? E a derivada $|g_2'(x)|$ perto da raiz?

### Tarefa 3: Newton
Use o método de `newton` para achar a raiz.
* Lembre-se: A derivada é $f'(x) = -e^{-x} - 1$.
* Compare o número de iterações do Newton com o da Bisseção.


# Tarefa 1:

In [50]:
f = lambda x: np.exp(-x) - x

raiz_bissecao = bissecao(f, 0, 1, tol=1e-6)

print(f"\nRaiz aproximada: {raiz_bissecao:.8f}")

Iter  a            b            x (raiz)     f(x)         Erro (Intervalo)
1     0.000000     1.000000     0.500000     1.07e-01     5.00e-01
2     0.500000     1.000000     0.750000     -2.78e-01    2.50e-01
3     0.500000     0.750000     0.625000     -8.97e-02    1.25e-01
4     0.500000     0.625000     0.562500     7.28e-03     6.25e-02
5     0.562500     0.625000     0.593750     -4.15e-02    3.12e-02
6     0.562500     0.593750     0.578125     -1.72e-02    1.56e-02
7     0.562500     0.578125     0.570312     -4.96e-03    7.81e-03
8     0.562500     0.570312     0.566406     1.16e-03     3.91e-03
9     0.566406     0.570312     0.568359     -1.91e-03    1.95e-03
10    0.566406     0.568359     0.567383     -3.75e-04    9.77e-04
11    0.566406     0.567383     0.566895     3.90e-04     4.88e-04
12    0.566895     0.567383     0.567139     7.24e-06     2.44e-04
13    0.567139     0.567383     0.567261     -1.84e-04    1.22e-04
14    0.567139     0.567261     0.567200     -8.84e-05

O número de iterações necessárias para atingir menos de 10e-6 de erro foi 20

# Tarefa 2:

In [51]:
def ponto_fixo(g, x0, tol=1e-6, max_iter=100):
    x = x0
    print(f"{'Iter':<5} {'x_old':<12} {'x_new':<12} {'Erro'}")

    for k in range(max_iter):
        x_new = g(x)

        erro = abs(x_new - x)
        print(f"{k+1:<5} {x:<12.6f} {x_new:<12.6f} {erro:.2e}")

        if erro < tol:
            return x_new

        x = x_new

    print("Número máximo de iterações atingido.")
    return x

In [52]:
# Forma A: g1(x) = e^(-x)
g1 = lambda x: np.exp(-x)
raiz_g1 = ponto_fixo(g1, x0=0.5)
print(f"Raiz (Forma A): {raiz_g1:.8f}")

Iter  x_old        x_new        Erro
1     0.500000     0.606531     1.07e-01
2     0.606531     0.545239     6.13e-02
3     0.545239     0.579703     3.45e-02
4     0.579703     0.560065     1.96e-02
5     0.560065     0.571172     1.11e-02
6     0.571172     0.564863     6.31e-03
7     0.564863     0.568438     3.58e-03
8     0.568438     0.566409     2.03e-03
9     0.566409     0.567560     1.15e-03
10    0.567560     0.566907     6.52e-04
11    0.566907     0.567277     3.70e-04
12    0.567277     0.567067     2.10e-04
13    0.567067     0.567186     1.19e-04
14    0.567186     0.567119     6.75e-05
15    0.567119     0.567157     3.83e-05
16    0.567157     0.567135     2.17e-05
17    0.567135     0.567148     1.23e-05
18    0.567148     0.567141     6.98e-06
19    0.567141     0.567145     3.96e-06
20    0.567145     0.567142     2.25e-06
21    0.567142     0.567144     1.27e-06
22    0.567144     0.567143     7.22e-07
Raiz (Forma A): 0.56714303


In [53]:
# Forma B: g2(x) = -ln(x)
g2 = lambda x: -np.log(x)
raiz_g2 = ponto_fixo(g2, x0=0.5, max_iter=10)

Iter  x_old        x_new        Erro
1     0.500000     0.693147     1.93e-01
2     0.693147     0.366513     3.27e-01
3     0.366513     1.003722     6.37e-01
4     1.003722     -0.003715    1.01e+00
5     -0.003715    nan          nan
6     nan          nan          nan
7     nan          nan          nan
8     nan          nan          nan
9     nan          nan          nan
10    nan          nan          nan
Número máximo de iterações atingido.


O método A funcionou e obteve o resultado esperado com erro menor que 10e-6 em 22 interações

Já o método B não obteve resultado significativo. Isso ocorre pois o método divirge.  
Calculando |g'(x)| perto da raiz (0.567), obtemos aproximadamente 1.76, o que explica o fato acima pela condição que |g'(x)| > 1.

# Tarefa 3

In [54]:
# Derivada f'(x) = -e^(-x) - 1
df = lambda x: -np.exp(-x) - 1

historico_newton = newton(f, df, x0=0.5, tol=1e-6)
raiz_newton = historico_newton[-1]
iteracoes_newton = len(historico_newton) - 1

print(f"\nRaiz aproximada (Newton): {raiz_newton.real:.8f}")
print(f"Iterações (Newton): {iteracoes_newton}")


Raiz aproximada (Newton): 0.56714314
Iterações (Newton): 2


O último elemento válido do histórico (antes do zero de preenchimento) é a raiz

A bisseção precisou de 20 iterações para reduzir intervalo inicial 1.0 a 1e-6.  
Já o método de Newton precisou de apenas 2 iterações.